### PR TITLE
[All] ObjectFactory: Explicit components registration 

### PR DESCRIPTION
**Pull request by fredroy**
- Task #3770 

Main goal is to remove the ObjectFactory singleton (later) ***and*** the implicit registration of objects while loading dynamic libraries (plugins)

In this PR, the explicit registration is applied on components of  `Sofa.Component.StateContainer`

The legacy RegisterObject mechanism is preserved and will warn the user at the compilation and at run-time.

EDIT:

- add tests about plugins non-registering/registering implicitly/registering explicitly
- add deprecations/warnings in RequiredPlugin
- add a function registerPlugin in PluginManager. This function is simply an alias for loadPlugin() because loadPlugin() does much more than just loading a plugin by itself.  In an eventual future, PluginManager should be refactored in a way that loadPlugin calls registerPlugin()

______________________________________________________

By submitting this pull request, I acknowledge that  
**I have read, understand, and agree [SOFA Developer Certificate of Origin (DCO)](https://github.com/sofa-framework/sofa/blob/master/CONTRIBUTING.md#sofa-developer-certificate-of-origin-dco)**.
______________________________________________________

**Reviewers will merge this pull-request only if**  
- it builds with SUCCESS for all platforms on the CI.
- it does not generate new warnings.
- it does not generate new unit test failures.
- it does not generate new scene test failures.
- it does not break API compatibility.
- it is more than 1 week old (or has fast-merge label).

### CODE REVIEW

File: Sofa/Component/StateContainer/src/sofa/component/statecontainer/init.cpp [hunk 1]
```diff
@@ -55,6 +60,39 @@ void init()
...
+
+        // Registration with RAII-style commit
+        // MechanicalObject
+        core::RegisterObject("mechanical state vectors", factory)
+            .add< MechanicalObject<Vec3Types> >(true) // default template
+            .add< MechanicalObject<Vec2Types> >()
+            .add< MechanicalObject<Vec1Types> >()
+            .add< MechanicalObject<Vec6Types> >()
+            .add< MechanicalObject<Rigid3Types> >()
+            .add< MechanicalObject<Rigid2Types> >();
```
Comment (alxbilger) [inline Sofa/Component/StateContainer/src/sofa/component/statecontainer/init.cpp]: I am thinking about the best design to register objects in the factory.

I think your proposition of code is error prone. You can forget to commit, or to add the factory as the second argument. It's probably due to the old design, and maybe we can think of a better design. Ideally, I would rather write something more explicit such as:

```cpp
const auto objectToRegister = core::RegisterObject(...).add<...>().add<...>();
factory->registerObject(objectToRegister);
```

And why not going further:

```cpp
factory->registerObject(core::RegisterObject("Mechanical state vectors in 2d").add<MechanicalObject<Vec2Types>>;
factory->registerObject(core::RegisterObject("Mechanical state vectors in 3d").add<MechanicalObject<Vec3Types>>;
...
```

Even the class `RegisterObject` bothers me. Nothing prevents not to call `add`. Maybe something like:

```cpp
template<class T>
class ObjectToRegister
{...};

factory->registerObject("Mechanical state vectors in 2d", ObjectToRegister<MechanicalObject<Vec2Types>>>;
```

or

```cpp
factory->registerObject<MechanicalObject<Vec2Types>>("Mechanical state vectors in 2d");
factory->registerObject<MechanicalObject<Vec3Types>>("Mechanical state vectors in 3d", ObjectFactory::defaultTemplate);

//ObjectFactory::defaultTemplate would be a strongly-typed boolean.
```

This way, you are obliged to use the variable `factory`, and no more a `RegisterObject` object which can be misleading, as it does not register object per se.

And if you want to keep the same description (you can always use a variable), you can write a parameter pack which would look like:

```cpp

class ObjectFactory
{
template<class ComponentType>
void registerObject();

template<class ComponentType, class RestOfComponentTypes...>
void registerObject()
{
    this->registerObject<ComponentType>();
    this->registerObject<RestOfComponentTypes...>();
}
};


factory->registerObject<
    MechanicalObject<Vec1Types>,
    MechanicalObject<Vec2Types>,
    DefaultTemplate<MechanicalObject<Vec3Types>>
>("Mechanical state vectors");
```

In summary, I would not continue to use a `RegisterObject` class. Note that I was thinking about interface only, not about the implementation which can be problematic, neither the compatibility with the old design.

File: Sofa/framework/Helper/src/sofa/helper/system/PluginManager.h [hunk 2]
```diff
@@ -141,6 +141,24 @@ class SOFA_HELPER_API Plugin
...
+        typedef typename LibraryEntry::FuncPtr FuncPtr;
+        entry.func = (FuncPtr)DynamicLibrary::getSymbolAddress(handle, entry.symbol);
+        if (entry.func == 0)
+        {
+            return false;
+        }
+        else
+        {
+            return true;
+        }
```
Comment (alxbilger) [inline Sofa/framework/Helper/src/sofa/helper/system/PluginManager.h]:
```suggestion
        return entry.func != 0;
```

File: Sofa/framework/Core/src/sofa/core/ObjectFactory.cpp [hunk 3]
```diff
@@ -635,29 +662,60 @@ RegisterObject::operator int()
...
+    void operator()(sofa::core::ObjectFactory* data)
+    {
+        if (func) return func(data);
+    }
+    RegisterObjectsEntry() :func(nullptr) {}
+} RegisterObjectsEntry;
+
+bool ObjectFactory::registerObjectsFromPlugin(const sofa::helper::system::Plugin& plugin)
+{
+    sofa::helper::system::PluginManager& pluginManager = sofa::helper::system::PluginManager::getInstance();
```
Comment (alxbilger) [inline Sofa/framework/Core/src/sofa/core/ObjectFactory.cpp:699]: In the quest of removing the singletons, you could pass the pluginManager instance as a parameter

**Review by alxbilger — COMMENTED** [review]
Very interesting PR. I have some suggestions to make the code clearer and more expressive.

**Comment by fredroy** [comment]
[ci-build][with-all-tests]

**Comment by fredroy** [comment]
[ci-build][with-all-tests][force-full-build]

**Comment by fredroy** [comment]
[ci-build][with-all-tests][force-full-build]

**Comment by damienmarchal** [comment]
Maybe some ideas here: https://github.com/sofa-framework/sofa/pull/2712

File: Sofa/Component/StateContainer/src/sofa/component/statecontainer/init.cpp [hunk 4]
```diff
@@ -51,10 +56,45 @@ void init()
...
+{
+    static bool registered = false;
+
+    if (!registered)
+    {
+        using namespace sofa::defaulttype;
+        
+        // Registration with explicit commit
+        // MappedObject
+        core::RegisterObject("Mapped state vectors")
```
Comment (damienmarchal) [inline Sofa/Component/StateContainer/src/sofa/component/statecontainer/init.cpp]: In init.cpp
```cpp
      void registerObjects(ObjectFactory* factory)
      {
            registerMappedObject(factory);
            ....
      }
      ....
```

which is a call to MappedObject.cpp and contains:
```cpp
void registerMappedObject(factory)
{
            factory->blabla->register
            .add< MappedObject<Vec1Types> >()
            .add< MappedObject<Vec3Types> >(true) // default template
            .add< MappedObject<Vec2Types> >()
            .add< MappedObject<Vec6Types> >()
            .add< MappedObject<Rigid3Types> >()
            .add< MappedObject<Rigid2Types> >()
            .commit(factory);
}
```

**Comment by fredroy** [comment]
[ci-build][with-all-tests][force-full-build]

**Comment by fredroy** [comment]
[ci-build][with-all-tests][force-full-build]

File: Sofa/framework/Core/src/sofa/core/ObjectFactory.h [hunk 5]
```diff
@@ -97,6 +107,10 @@ class SOFA_CORE_API ObjectFactory
     ClassEntryMap registry;
     OnCreateCallback m_callbackOnCreate ;
 
+    /// Keep track of plugins who already registered
+    using RegisteredPluginMap = std::map<std::string, bool>;
```
Comment (alxbilger) [inline Sofa/framework/Core/src/sofa/core/ObjectFactory.h]: This is a simply a `std::set<std::string>`

File: Sofa/Component/src/sofa/component/init.cpp [hunk 6]
```diff
@@ -68,6 +72,11 @@ const char* getModuleVersion()
     return MODULE_VERSION;
 }
 
+void registerObjects(sofa::core::ObjectFactory* factory)
+{
+    factory->registerObjectsFromPlugin("Sofa.Component.StateContainer");
```
Comment (alxbilger) [inline Sofa/Component/src/sofa/component/init.cpp:76]: I am annoyed that a meta-module must call the "dynamic" functions (using strings) as if it does not have access to it statically.

File: Sofa/framework/Core/src/sofa/core/ObjectFactory.cpp [hunk 7]
```diff
@@ -635,29 +641,117 @@ RegisterObject::operator int()
...
+    typedef void (*FuncPtr) (sofa::core::ObjectFactory*);
+    FuncPtr func;
+    void operator()(sofa::core::ObjectFactory* data)
+    {
+        if (func) return func(data);
+    }
+    ObjectRegistrationEntry() :func(nullptr) {}
+} ObjectRegistrationEntry;
+
+bool ObjectFactory::registerObjectsFromPlugin(const std::string& pluginName)
```
Comment (alxbilger) [inline Sofa/framework/Core/src/sofa/core/ObjectFactory.cpp:697]: Instead, I suggest the following 
```suggestion
bool ObjectFactory::registerObjectsFromPlugin(const helper::system::Plugin& plugin)
```

In `Main.cpp` , you would need to do:

```cpp
    for (const auto& [_, plugin] : pluginManager.getPluginMap())
    {
        objectFactory->registerObjectsFromPlugin(plugin);
    }
```

Comment (fredroy) [reply Sofa/framework/Core/src/sofa/core/ObjectFactory.cpp:697]: Actually it was what it was done at the beginning https://github.com/sofa-framework/sofa/blob/c36c1458c968d59275eec5e69b219a9bc8177de8/Sofa/framework/Core/src/sofa/core/ObjectFactory.h#L198

But how do use it in [Sofa/Component/src/sofa/component/init.cpp](https://github.com/sofa-framework/sofa/pull/4429/files/6eacd8030a1f4e532b0ce705464616fc71ef971b#diff-1dd06bf92fc1ac386c2d549a88e5072dbbaa36714904ac62d448f36b531ffebd) ? You would still need to:
 - get the pluginmanager (of course)
 - find the plugin in the pluginMap (with its string ID)
 - then calling registerObjectsFromPlugin on the retrieved pointer (after a nullptr check)

Comment (alxbilger) [reply Sofa/framework/Core/src/sofa/core/ObjectFactory.cpp:697]: I understand. But this is what bothers me. I don't like that a meta-module needs the `PluginManager` (and even a `Plugin` object) to be able to interact with its own dependencies. If we use the `PluginManager` in `Sofa.Component` to load all the sub-modules, it's not even necessary to have them as a dependency of the module.

File: Sofa/Component/src/sofa/component/init.cpp (hunk 6 above)
Comment (damienmarchal) [reply Sofa/Component/src/sofa/component/init.cpp:76]:
```cpp
sofa::component::statecontainer::target = "Sofa.Component.StateContainer"
```

**Comment by fredroy** [comment]
Closing (postponing), several issues detected in this PR
- PluginManager is not satisfactory
- meta-modules are not well designed so induced too much complexities

**Comment by fredroy** [comment]
[ci-build][with-all-tests][force-full-build]

**Comment by fredroy** [comment]
[ci-build][with-all-tests]

**Comment by fredroy** [comment]
[ci-build][with-all-tests][force-full-build]
